File: backend/src/wind_track/models/schemas.py
"""Pydantic API schemas."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class HealthResponse(BaseModel):
    status: str
    version: str


class AreaResponse(BaseModel):
    id: int
    slug: str
    name: str
    area_type: str
    center_lat: float
    center_lon: float
    default_zoom: float
    active: bool


class DataQualityResponse(BaseModel):
    area_id: int
    building_count: int
    official_height_coverage: float
    estimated_height_coverage: float
    fallback_height_coverage: float
    missing_height_count: int
    roads_with_inferred_width: int
    vegetation_count: int
    special_geometry_counts: dict[str, int]
    low_confidence_count: int


class ScalarScenarioRequest(BaseModel):
    area_slug: str = "pilot_presquile"
    wind_speed_ms: float = Field(ge=0, le=40)
    wind_direction_deg: float = Field(ge=0, lt=360)
    scenario_type: Literal["manual", "current_weather", "forecast"] = "manual"
    weather_observation_id: int | None = None
    wind_gust_ms: float | None = None


class ScenarioResponse(BaseModel):
    scenario_id: int
    area_slug: str
    wind_speed_ms: float
    wind_direction_deg: float
    scenario_type: str
    feature_count: int
    model_version: str
    data_version: str


class FeatureResultResponse(BaseModel):
    feature_id: int
    feature_type: str
    name: str | None
    subtype: str | None = None
    geom: dict[str, Any]
    risk_score: float
    exposure_class: str
    local_multiplier: float
    approx_local_speed_ms: float | None
    gust_sensitive: bool
    confidence: float
    handling_mode: str
    subscores: dict[str, Any]
    cause_tags: list[str]
    mitigation_tags: list[str]
    model_note: str | None
    limitations: list[str]
    cache_hit: bool | None = None
    cache_direction_deg: float | None = None


class WeatherResponse(BaseModel):
    area_id: int
    wind_speed_10m_ms: float | None
    wind_direction_10m_deg: float | None
    wind_gust_10m_ms: float | None
    timestamp: str
    source: str
    is_forecast: bool = False
    forecast_timestamp: str | None = None


class FeedbackRequest(BaseModel):
    area_id: int
    feature_id: int | None = None
    feedback_type: str
    description: str
    wind_direction_deg: float | None = None
    weather_context: dict[str, Any] = Field(default_factory=dict)


class FeedbackResponse(BaseModel):
    id: int
    status: str


class ValidationCaseResponse(BaseModel):
    id: int
    name: str
    case_type: str
    area_id: int
    wind_direction_deg: float | None
    reference_speed_ms: float | None
    sample_count: int = 0


class ValidationRunRequest(BaseModel):
    validation_case_id: int
    model_version_id: int | None = None


class ValidationMetricsResponse(BaseModel):
    validation_case_id: int
    overall_accuracy: float
    high_wind_recall: float | None
    high_wind_precision: float | None
    adjacent_class_accuracy: float
    false_negative_rate: float | None
    metrics_json: dict[str, Any]

File: frontend/src/api/schemas.ts
import { z } from "zod";

export const HealthSchema = z.object({
  status: z.string(),
  version: z.string(),
}).strict();

export const AreaSchema = z.object({
  id: z.number(),
  slug: z.string(),
  name: z.string(),
  area_type: z.string(),
  center_lat: z.number(),
  center_lon: z.number(),
  default_zoom: z.number(),
  active: z.boolean(),
}).strict();

export const ScenarioRequestSchema = z.object({
  area_slug: z.string(),
  wind_speed_ms: z.number(),
  wind_direction_deg: z.number(),
  scenario_type: z.enum(["manual", "current_weather", "forecast"]),
  weather_observation_id: z.number().optional(),
  wind_gust_ms: z.number().optional(),
}).strict();

export const ScenarioSchema = z.object({
  scenario_id: z.number(),
  area_slug: z.string(),
  wind_speed_ms: z.number(),
  wind_direction_deg: z.number(),
  scenario_type: z.string(),
  feature_count: z.number(),
  model_version: z.string(),
  data_version: z.string(),
}).strict();

export const FeatureResultSchema = z.object({
  feature_id: z.number(),
  feature_type: z.string(),
  name: z.string().nullable(),
  subtype: z.string().nullable().optional(),
  geom: z.record(z.string(), z.unknown()),
  risk_score: z.number(),
  exposure_class: z.string(),
  local_multiplier: z.number(),
  approx_local_speed_ms: z.number().nullable(),
  gust_sensitive: z.boolean(),
  confidence: z.number(),
  handling_mode: z.string(),
  subscores: z.record(z.string(), z.unknown()),
  cause_tags: z.array(z.string()),
  mitigation_tags: z.array(z.string()),
  model_note: z.string().nullable(),
  limitations: z.array(z.string()),
  cache_hit: z.boolean().nullish(),
  cache_direction_deg: z.number().nullish(),
}).strict();

export const WeatherSchema = z.object({
  area_id: z.number(),
  wind_speed_10m_ms: z.number().nullable(),
  wind_direction_10m_deg: z.number().nullable(),
  wind_gust_10m_ms: z.number().nullable(),
  timestamp: z.string(),
  source: z.string(),
  is_forecast: z.boolean().optional(),
  forecast_timestamp: z.string().nullable().optional(),
}).strict();

export const TileManifestSchema = z.object({
  area_slug: z.string(),
  ready: z.boolean(),
  base_pmtiles: z.boolean(),
  exposure_pmtiles: z.record(z.string(), z.boolean()),
  tippecanoe_available: z.boolean(),
  tiles_path: z.string(),
}).strict();

export const DataQualitySchema = z.object({
  area_id: z.number(),
  building_count: z.number(),
  official_height_coverage: z.number(),
  estimated_height_coverage: z.number(),
  fallback_height_coverage: z.number(),
  missing_height_count: z.number(),
  roads_with_inferred_width: z.number(),
  vegetation_count: z.number(),
  special_geometry_counts: z.record(z.string(), z.number()),
  low_confidence_count: z.number(),
}).strict();

export type TileManifest = z.infer<typeof TileManifestSchema>;
export type Area = z.infer<typeof AreaSchema>;
export type DataQuality = z.infer<typeof DataQualitySchema>;
export type Scenario = z.infer<typeof ScenarioSchema>;
export type FeatureResult = z.infer<typeof FeatureResultSchema>;
export type Weather = z.infer<typeof WeatherSchema>;